lib/text/mention_text_editing_controller.dart

```diff
@@ -1,121 +1,76 @@
-  final String avatarUrl;
-}
-
-// Mention syntax for determining when to start mentioning and parsing to and from markup
-// Final markup text would be Prefix -> StartingCharacter -> Id of mention -> Suffix
-class MentionSyntax {
-  MentionSyntax(
-      {required this.startingCharacter,
-      required this.missingText,
-      this.prefix = '<###',
-      this.suffix = '###>',
-      this.pattern = "[a-zA-Z0-9]{1,}"}) {
-    _mentionRegex = RegExp('($prefix)($startingCharacter)($pattern)($suffix)');
-  }
-
-  // The character the regex pattern starts with, used to more performantly find sections in the text, needs to be a single character
-  final String startingCharacter;
-
-  // The prefix to add to the final markup text per mention of this type
-  final String prefix;
-
-  // The suffix to add to the final markup text per mention of this type
-  final String suffix;
-
-  // The display name to show when the mention with the specified id no longer exists
-  final String missingText;
-
-  // The inner pattern that will be followed to find a mention
-  final String pattern;
-
-  late RegExp _mentionRegex;
+/// Typedef for suggestion changed callback.
+typedef SuggestionChangedCallback = void Function(
+  MentionSyntax? syntax,
+  String? searchString,
+);
 
-  RegExp getRegExp() => _mentionRegex;
-}
-
-// Local-only class to store mentions currently stored in the string visible to the user
-class _TextMention {
-  _TextMention(
-      {required this.id,
-      required this.display,
-      required this.start,
-      required this.end,
-      required this.syntax});
-
-  final String id;
-  final String display;
-  final MentionSyntax syntax;
-  int start;
-  int end;
-}
+/// Typedef for id to mention callback.
+typedef IdToMentionCallback = MentionObject? Function(
+  BuildContext context,
+  String mentionId,
+);
 
-// Text editing controller that can parse mentions
+/// Text editing controller that can parse mentions
 class MentionTextEditingController extends TextEditingController {
   MentionTextEditingController({
     this.controllerToCopyTo,
     required this.mentionSyntaxes,
-    this.onSugggestionChanged,
-    required this.mentionBgColor,
-    required this.mentionTextColor,
-    required this.mentionTextStyle,
-    required this.runTextStyle,
+    this.onSuggestionChanged,
+    this.mentionBgColor,
+    this.mentionTextColor,
+    this.mentionTextStyle,
+    this.runTextStyle,
     required this.idToMentionObject,
     super.text,
   }) {
     _init();
   }
 
-  // Unique mention syntaxes, all syntaxes should have a different starting character
+  /// Unique mention syntaxes, all syntaxes should have a different
+  /// starting character
   final List<MentionSyntax> mentionSyntaxes;
 
-  // Delegate called when suggestion has changed
-  Function(MentionSyntax? syntax, String?)? onSugggestionChanged;
+  /// Delegate called when suggestion has changed
+  final SuggestionChangedCallback? onSuggestionChanged;
```

Comment:
This property was intentionally not final as we have use cases where we need to bind to the callback after the controller was created. Can you remove the final here? Once merged we'll then add final back, but instead use functions to add and remove listeners to this callback.

Reply:
implemented listeners